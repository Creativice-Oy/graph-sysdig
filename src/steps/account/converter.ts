import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';
import { SysdigAccount } from '../../types';

import { Entities } from '../constants';

export function getAccountKey(id: number): string {
  return `sysdig_account:${id}`;
}

export function createAccountEntity(data: SysdigAccount): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getAccountKey(data.id as number),
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        id: (data.id as number).toString(),
        name: data.name,
        displayName: `${data.firstName} ${data.lastName}`,
        username: data.username,
      },
    },
  });
}
