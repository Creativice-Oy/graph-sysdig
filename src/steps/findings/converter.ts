import {
  createIntegrationEntity,
  createMappedRelationship,
  Entity,
  parseTimePropertyValue,
  Relationship,
  RelationshipClass,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import { Vulnerability, VulnerablePackage } from '../../types';

import { Entities, Relationships } from '../constants';

export function getFindingKey(id: string): string {
  return `sysdig_finding:${id}`;
}

function getSeverity(score: number): string {
  switch (score) {
    case 7:
      return 'Negligible';
    case 6:
      return 'Low';
    case 5:
      return 'Medium';
    case 3:
      return 'High';
    case 2:
      return 'Critical';
    default:
      return 'Unknown';
  }
}

export function createFindingEntity(data: VulnerablePackage): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getFindingKey(data.id),
        _type: Entities.FINDING._type,
        _class: Entities.FINDING._class,
        id: data.id,
        name: data.vuln.name,
        category: 'Package',
        severity: getSeverity(data.vuln.severity.value),
        numericSeverity: data.vuln.severity.value,
        open: true,
        'vulnerability.name': data.vuln.name,
        'vulnerability.severity.reporter.name':
          data.vuln.severity.reporter.name,
        'vulnerability.severity.reporter.url': data.vuln.severity.reporter.url,
        'vulnerability.severity.value': data.vuln.severity.value,
        'vulnerability.cvssScore.reporter.name':
          data.vuln.cvssScore.reporter.name,
        'vulnerability.cvssScore.reporter.url':
          data.vuln.cvssScore.reporter.url,
        'vulnerability.cvssScore.value.version':
          data.vuln.cvssScore.value.version,
        'vulnerability.cvssScore.value.score': data.vuln.cvssScore.value.score,
        'vulnerability.cvssScore.value.vector':
          data.vuln.cvssScore.value.vector,
        'vulnerability.exploitable': data.vuln.exploitable,
        'vulnerability.exploit': data.vuln.exploit,
        'vulnerability.description': data.vuln.description,
        'vulnerability.disclosureDate': parseTimePropertyValue(
          data.vuln.disclosureDate,
        ),
        'vulnerability.solutionDate': parseTimePropertyValue(
          data.vuln.solutionDate,
        ),
        'package.id': data.package.id,
        'package.name': data.package.name,
        'package.version': data.package.version,
        'package.type': data.package.type,
        'package.path': data.package.path,
        'package.running': data.package.running,
        fixedInVersion: data.vuln.name,
      },
    },
  });
}

export function createFindingCveRelationship(
  finding: Entity,
  cve: Vulnerability,
): Relationship {
  const disclosureDate = parseTimePropertyValue(cve.disclosureDate);
  const solutionDate = parseTimePropertyValue(cve.solutionDate);

  return createMappedRelationship({
    _class: RelationshipClass.IS,
    _type: Relationships.FINDING_IS_CVE._type,
    _mapping: {
      sourceEntityKey: finding._key,
      relationshipDirection: RelationshipDirection.FORWARD,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', '_key']],
      targetEntity: {
        _key: cve.name.toLowerCase(),
        _type: Entities.CVE._type,
        id: cve.name,
        score: cve.cvssScore.value.score,
        name: cve.name,
        displayName: cve.name,
        'severity.reporter.name': cve.severity.reporter.name,
        'severity.reporter.url': cve.severity.reporter.url,
        'severity.value': cve.severity.value,
        'cvssScore.reporter.name': cve.cvssScore.reporter.name,
        'cvssScore.reporter.url': cve.cvssScore.reporter.url,
        'cvssScore.value.version': cve.cvssScore.value.version,
        'cvssScore.value.score': cve.cvssScore.value.score,
        'cvssScore.value.vector': cve.cvssScore.value.vector,
        exploitable: cve.exploitable,
        exploit: cve.exploit,
        description: cve.description,
        disclosureDate: disclosureDate,
        solutionDate: solutionDate,
      },
    },
  });
}
