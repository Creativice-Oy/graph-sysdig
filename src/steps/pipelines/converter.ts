import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { SysdigPipeline } from '../../types';

import { Entities } from '../constants';

export function getPipelineKey(id: string): string {
  return `sysdig_image_scan:${id}`;
}

export function createPipelineEntity(pipeline: SysdigPipeline): Entity {
  return createIntegrationEntity({
    entityData: {
      source: pipeline,
      assign: {
        _key: getPipelineKey(pipeline.id),
        _type: Entities.PIPELINE._type,
        _class: Entities.PIPELINE._class,
        id: pipeline.id,
        storedAt: parseTimePropertyValue(pipeline.storedAt),
        imageId: pipeline.imageId,
        name: pipeline.imagePullString,
        imagePullString: pipeline.imagePullString,
        vulnsBySev: pipeline.vulnsBySev,
        exploitCount: pipeline.exploitCount,
        policyEvaluationsPassed: pipeline.policyEvaluationsPassed,
      },
    },
  });
}
