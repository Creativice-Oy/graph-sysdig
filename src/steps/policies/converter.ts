import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Policy } from '../../types';

import { Entities } from '../constants';

export function getPolicyKey(identifier: string): string {
  return `sysdig_policy:${identifier}`;
}

export function createPolicyEntity(data: Policy): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getPolicyKey(data.identifier),
        _type: Entities.POLICY._type,
        _class: Entities.POLICY._class,
        id: data.id.toString(),
        name: data.name,
        title: data.name,
        identifier: data.identifier,
        summary: data.description,
        content: data.description,
        creationTimestamp: parseTimePropertyValue(data.creationTimestamp),
        updateTimestamp: parseTimePropertyValue(data.updateTimestamp),
      },
    },
  });
}
