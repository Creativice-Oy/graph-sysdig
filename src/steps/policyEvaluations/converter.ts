import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { PolicyEvaluation } from '../../types';
import { Entities } from '../constants';

export function getPolicyEvaluationKey(id: string): string {
  return `sysdig_finding:${id}`;
}

export function createPolicyEvaluationEntity(
  data: PolicyEvaluation,
  imageScan: Entity,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getPolicyEvaluationKey(data.id),
        _type: Entities.POLICY_EVALUATION._type,
        _class: Entities.POLICY_EVALUATION._class,
        id: data.id,
        name: data.name,
        category: 'Vulnerability Scan',
        summary: data.name,
        internal: true,
        completedOn: imageScan.storedAt ?? 0,
        identifier: data.identifier,
        policyType: data.policyType,
        evaluationResult: data.evaluationResult,
        'failuresCount.imageConfigCreationDate':
          data.failuresCount.imageConfigCreationDate,
        'failuresCount.imageConfigDefaultUser':
          data.failuresCount.imageConfigDefaultUser,
        'failuresCount.imageConfigEnvVariable':
          data.failuresCount.imageConfigEnvVariable,
        'failuresCount.imageConfigInstructionIsPkgManager':
          data.failuresCount.imageConfigInstructionIsPkgManager,
        'failuresCount.imageConfigLabel': data.failuresCount.imageConfigLabel,
        'failuresCount.imageConfigSensitiveInformationAndSecrets':
          data.failuresCount.imageConfigSensitiveInformationAndSecrets,
        'failuresCount.vulnDenyList': data.failuresCount.vulnDenyList,
        'failuresCount.imageConfigInstructionNotRecommended':
          data.failuresCount.imageConfigInstructionNotRecommended,
        'failuresCount.vulnSeverityAndThreats':
          data.failuresCount.vulnSeverityAndThreats,
        creationTimestamp: parseTimePropertyValue(data.creationTimestamp),
        updateTimestamp: parseTimePropertyValue(data.updateTimestamp),
      },
    },
  });
}
