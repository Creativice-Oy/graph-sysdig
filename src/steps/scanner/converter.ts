import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';
import { SysdigAccount } from '../../types';

import { Entities } from '../constants';

export function getScannerKey(id: number): string {
  return `sysdig_scanner:${id}`;
}

export function createScannerEntity(data: SysdigAccount): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getScannerKey(data.id as number),
        _type: Entities.SCANNER._type,
        _class: Entities.SCANNER._class,
        id: (data.id as number).toString(),
        name: data.name,
        displayName: `${data.firstName} ${data.lastName}`,
        username: data.username,
        version: data.version,
        products: data.products,
        category: ['image'],
        function: ['scanning'],
      },
    },
  });
}
