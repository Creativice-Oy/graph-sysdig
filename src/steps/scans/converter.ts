import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { ImageScanV2, SysdigResult } from '../../types';
import { Entities } from '../constants';

export function getImageScanKey(id: string): string {
  return `sysdig_image_scan:${id}`;
}

export function createImageScanEntity(data: SysdigResult): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {
        ...data,
        createdOn: data.createdAt,
      },
      assign: {
        _key: getImageScanKey(data.imageId),
        _type: Entities.IMAGE_SCAN._type,
        _class: Entities.IMAGE_SCAN._class,
        name: data.repository,
        analysisStatus: data.analysisStatus,
        analyzedAt: data.analyzedAt,
        createdAt: data.createdAt,
        fullTag: data.fullTag,
        imageDigest: data.imageDigest,
        imageId: data.imageId,
        parentDigest: data.parentDigest,
        tagDetectedAt: data.tagDetectedAt,
        registry: data.registry,
        repository: data.repository,
        tag: data.tag,
        origin: data.origin,
        policyStatus: data.policyStatus,
        category: 'Image Scan',
        summary: 'Image Scan',
        internal: false,
      },
    },
  });
}

export function createImageScanEntityV2(data: ImageScanV2): Entity {
  const storedAt = parseTimePropertyValue(data.storedAt);
  const createdAt = parseTimePropertyValue(data.metadata.createdAt);
  const vulnsBySev = data.vulnsBySev?.map((vuln) => vuln.fixableCount ?? 0);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getImageScanKey(data.id),
        _type: Entities.IMAGE_SCAN._type,
        _class: Entities.IMAGE_SCAN._class,
        id: data.id,
        name: data.metadata.pullString,
        storedAt,
        type: data.type,
        'metadata.imageId': data.metadata.imageId,
        'metadata.pullString': data.metadata.pullString,
        'metadata.baseOS': data.metadata.baseOS,
        'metadata.digest': data.metadata.digest,
        'metadata.createdAt': createdAt,
        'metadata.author': data.metadata.author,
        'metadata.size': data.metadata.size,
        'metadata.os': data.metadata.os,
        'metadata.architecture': data.metadata.architecture,
        'metadata.labels': data.metadata.labels,
        'metadata.layersCount': data.metadata.layersCount,
        vulnsBySev,
        packageCount: data.packageCount,
        packageTypes: data.packageTypes,
        fixablePackages: data.fixablePackages,
        runningFixablePackages: data.runningFixablePackages,
        category: 'Image Scan',
        summary: 'Image Scan',
        internal: false,
      },
    },
  });
}
