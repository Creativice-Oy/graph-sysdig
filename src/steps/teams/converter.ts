import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';
import { SysdigTeam } from '../../types';

import { Entities } from '../constants';

export function getTeamKey(id: number): string {
  return `sysdig_team:${id}`;
}

export function createTeamEntity(data: SysdigTeam): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getTeamKey(data.id as number),
        _type: Entities.TEAM._type,
        _class: Entities.TEAM._class,
        id: data.id.toString(),
        version: data.version,
        createdOn: data.dateCreated,
        updatedOn: data.lastUpdated,
        customerId: data.customerId,
        immutable: data.immutable,
        name: data.name,
        theme: data.theme,
        description: data.description,
        show: data.show,
        origin: data.origin,
        canUseSysdigCapture: data.canUseSysdigCapture,
        canUseAgentCli: data.canUseAgentCli,
        canUseCustomEvents: data.canUseCustomEvents,
        canUseAwsMetrics: data.canUseAwsMetrics,
        canUseBeaconMetrics: data.canUseBeaconMetrics,
        canUseRapidResponse: data.canUseRapidResponse,
        defaultTeamRole: data.defaultTeamRole,
        userCount: data.userCount,
        entryPointModule: data.entryPoint.module,
        products: data.products,
        default: data.default,
      },
    },
  });
}
