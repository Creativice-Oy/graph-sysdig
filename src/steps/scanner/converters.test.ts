import { createScannerEntity } from './converter';

describe('#createScannerEntity', () => {
  test('should convert to entity', () => {
    expect(createScannerEntity({ name: 'Sysdig Scanner' })).toMatchSnapshot();
  });
});
