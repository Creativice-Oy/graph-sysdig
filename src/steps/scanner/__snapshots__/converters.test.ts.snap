// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createScannerEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "Service",
  ],
  "_key": "sysdig_scanner:Sysdig Scanner",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "name": "Sysdig Scanner",
      },
    },
  ],
  "_type": "sysdig_scanner",
  "category": Array [
    "image",
  ],
  "createdOn": undefined,
  "displayName": "Sysdig Scanner",
  "function": Array [
    "scanning",
  ],
  "name": "Sysdig Scanner",
}
`;
